Dylan tong aws blog 2023 opensearch ai retrospective (#2516)
opensearch-project/project-website: _posts/2024-01-02-opensearch-ai-retrospective.md

---
layout: post
title:  "OpenSearch AI and machine learning: Recapping 2023 and what’s next"
authors:
 - dylatong
date: 2024-01-02
categories:
 - community-updates, technical-posts, roadmap

> meta_keywords: vector database, vector search, ai, artificial intelligence, machine learning, neural-search, ml-commons, k-nn, ml-framework

Reviewer: Please update the meta keywords and description with: 

**meta_keywords:** OpenSearch strategic enhancements, OpenSearch core engine, vector search engine, OpenSearch GenAI, machine learning framework

**meta_description:** Learn about strategic enhancements to the OpenSearch core engine since the Elasticsearch 7.10.2. fork and how they have improved the performance of log analytics and search use cases.

![flow-designer](/assets/media/blog-images/2024-01-02-opensearch-ai-retrospective/flow-designer.png)*Build AI apps faster by composing AI flows through OpenSearch’s no-code experience*([RFC](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4755))

### **New AI Use Case Support**

With the composable flow paradigm, we envision supporting the existing use cases, variations and a plethora of new possibilities. Among the possible use cases will be the ability to construct re-ranking workflows. You will be able to compose a workflow that can retrieve results through lexical or vector methods and efficiently send those results to a re-ranking model, which might be a custom model or a provider like Cohere ReRank. These workflows will unlock various use cases, including search relevancy boosting through state-of-the-art techniques like combining bi-encoder and cross-encoder models and user-level personalization by re-ranking results using deep recommender models. 

> As well, the ability to use the generic ML processor within last-mile ingestion pipelines, will unlock countless use cases based around automating metadata extraction to enrich indexes. We plan to provide you with an ingest processor that allows you to configure your model interface, so that you can more easily integrate models ranging from name-entity-recognition models, computer vision classifiers, sentiment detectors to neural machine translators to enrich your data during ingestion and indexing flows.

Reviewer 2: ```suggestion
Additionally, the ability to use the generic ML processor within last-mile ingest pipelines will unlock countless use cases related to automating metadata extraction in order to enrich indexes. We plan to provide you with an ingest processor that allows you to configure your model interface so that you can more easily integrate models, including named-entity recognition models, computer vision classifiers, sentiment detectors, and neural machine translators, to enrich your data during ingestion and indexing flows.
```

> ### **Machine Learning Framework Enhancements**

Reviewer 2: ```suggestion
### ML framework enhancements
```

> 

Reviewer 2: Text should separate these headings so that they are not "stacked." Something like "The following subsections provide information about..." is usually fine.

> #### ML Extensibility Enhancements

Reviewer 2: ```suggestion
#### ML extensibility enhancements
```

#### Batch Optimized Inference

Among the things we plan to enhance next year is augmenting our framework so that our AI connectors have the ability to integrate with batch optimized APIs and services. We want you to be be able to easily enrich your ingest pipelines with a variety of ML models to generate vector embeddings and metadata. These types of ingest pipelines are typically best ran as a batch process, so we want you to have access to a variety of connectors that can deliver optimal price performance on batch workloads. All the while, we want you to have flexibility and the freedom to choose your preferred tooling.

#### **Certified AI Connectors**

Additionally, we plan to implement a process for certifying connectors to maintain a quality standard for our users. We plan to elevate certified connectors by making them more accessible, for instance, by including them in our distributions by default.

> Connector authors that acquire certification will qualify for additional benefits. For instance, they could qualify for publishing their connectors in the Amazon OpenSearch Services integration catalog, which enable AWS customers to automate infrastructure and OpenSearch resource creation to install connectors on managed clusters. Currently, to install a connector, you have to copy connector blueprints from our documentation. Then, you have to manually create OpenSearch and AWS resources like OpenSearch model abstractions and IAM policies. These automations make it easier for customers to use our AI integrations, and certified connector authors will have privileged access to mechanisms that facilitate go-to-market objectives.

Reviewer 2: Is the name of the catalog actually "Amazon OpenSearch Services"?

![forecasting](/assets/media/blog-images/2024-01-02-opensearch-ai-retrospective/forecasting.png)*Forecasting cpu usage across a global server fleet*([RFC](https://github.com/opensearch-project/ux/issues/70))

### Shape the Future of OpenSearch AI and Machine Learning

We hope you're excited about what's to come for OpenSearch AI and Machine Learning in 2024. As always, we're interested in your feedback about our roadmap. You can share your thoughts with us in the [forum](https://forum.opensearch.org/t/feedback-whats-on-your-wishlist-for-opensearch-ai-and-machine-learning-capabilities-in-2024/17099), or through our community [feature request forms](https://github.com/opensearch-project/.github/issues/new/choose) to help us prioritize our efforts.

Additionally, if you like to learn more about the 2023 AI/ML features that were highlighted in this blog, the following are some technical blog posts that we published on these topics:

> * [Using and building OpenSearch AI connectors](https://community.aws/content/2ZUiRDEKnIg0PiwPevVTaJ7B9qg/introduction-to-opensearch-models)

Reviewer 2: ```suggestion
* [Introduction to OpenSearch Models](https://community.aws/content/2ZUiRDEKnIg0PiwPevVTaJ7B9qg/introduction-to-opensearch-models)
```

* [Efficient filtering for vector search](https://opensearch.org/blog/efficient-filters-in-knn/#:~:text=Efficient%20filters%20are%20supported%20in,and%20were%20called%20Lucene%20Filters.)
* [Byte-quantized vector support](https://opensearch.org/blog/byte-quantized-vectors-in-opensearch/)
* [Hybrid search score normalization](https://opensearch.org/blog/hybrid-search/)
* [Sparse encoding based semantic search](https://opensearch.org/blog/improving-document-retrieval-with-sparse-semantic-encoders/)

> * Multi-modal search for neural search (in progress)

Author: yeah, I was hoping we could get the MM blog published before this one. It's not going to happen. I like to add a link to the MM blog post publication.
